File: binarysearch.com/WC38/4_VirtualArray.cpp
class VirtualArray
{
    map<long long, long long> intervals;

public:
    VirtualArray()
    {
        intervals.insert({LLONG_MIN, LLONG_MIN});
        intervals.insert({LLONG_MAX, LLONG_MAX});
    }

    void set(int start, int end)
    {
        long long s = start;
        long long e = end;

        auto ptr = intervals.upper_bound(s);
        ptr--;
        if (ptr->second >= s)
        {
            s = min(s, ptr->first);
            e = max(e, ptr->second);
            intervals.erase(ptr->first);
        }
        long long next_start = s;
        while (intervals.upper_bound(next_start) != intervals.end())
        {
            auto pair = intervals.upper_bound(next_start);
            if (pair->first > e)
                break;

            next_start = pair->first;
            e = max(e, pair->second);
            intervals.erase(pair->first);
        }
        intervals.insert({s, e});
    }

    bool get(int idx)
    {
        auto ptr = intervals.upper_bound(idx);
        ptr--;
        return ptr->second >= idx;
    }
};

File: binarysearch.com/WC38/4_VirtualArray.java
import java.util.*;

class VirtualArray {
    TreeMap<Integer, Integer> intervals;

    public VirtualArray() {
        this.intervals = new TreeMap();
    }

    public void set(int start, int end) {
        Integer floorkey = intervals.floorKey(start);
        if (floorkey != null && intervals.get(floorkey) >= end) {
            start = Math.min(start, floorkey);
            end = Math.max(end, intervals.get(floorkey));
            intervals.remove(floorkey);
        }

        int merge_start = start;

        // merge intervals
        while (intervals.ceilingKey(merge_start) != null) {
            int k = intervals.ceilingKey(merge_start);
            if (k > end)
                break;
            end = Math.max(end, intervals.get(k));
            intervals.remove(k);
            merge_start = k;
        }

        intervals.put(start, end);
    }

    public boolean get(int idx) {
        Integer key = intervals.floorKey(idx);
        if (key == null)
            return false;

        return intervals.get(key) >= idx;
    }
}